doc/erofs.md

```diff
@@ -0,0 +1,431 @@
+the superblock.
+
+This feature got added in kernel commits:
+*   `3f339920175c` ("erofs: update on-disk format for xattr name filter")
+*   `fd73a4395d47` ("erofs: boost negative xattr lookup with bloom filter")
+
+### Data layout
+
+erofs has a bunch of different ways to represent the actual content associated
+with an inode (regular file content, directory entries, symlink target).
+
+We describe three of them here:
+*   plain
+*   inline
+*   chunked
+
+The data layout is chosen using some of the bits of the `format` field in the
+inode header.
+
+#### `EROFS_INODE_FLAT_PLAIN`
+
+In this case there's never any inline data.  The inode content is stored
+entirely as a series of contiguous blocks.  The offset of the first block is
+what goes in the `i_u` field (measured in blocks, not bytes).
+
+The number of blocks is determined by the `.size` field (divided by block size,
+rounded up).
+
+If the content is not a multiple of the blocksize then the last block should be
+0-padded.
+
+#### `EROFS_INODE_FLAT_INLINE`
+
+This is similar to `EROFS_INODE_FLAT_PLAIN` except if the content is not a
+multiple of the blocksize.  In that case, instead of 0-padding the last block
+to fill up a block, the content of the last block is stored directly inline
+with the inode, without padding.
+
+So, imagining the content is 2.5 blocks worth of data:
+*   the first block is the one pointed to by `i_u`
+*   the second block is the one immediately following it
+*   the last block is stored at the end of the inode
+
+The number of blocks is determined by the `.size` field, divided by block size,
+rounded down.  The remainder is the number of bytes of inline data.
+
+The inline data must be written in such a way that it does not cross a block
+boundary.  It is theoretically permitted for the inline data to be in a
+separate block (ie: the block directly following the inode data).  It is also
+permitted for the inode data itself to cross block boundaries.  There are a
+couple of caveats to be aware of, however:
+*   the alignment of inodes is 32 bytes, but the size of an extended inode is 64
+    bytes.  `mkfs.erofs` tries to ensure that extended inodes headers land
+    entirely within on disk block (for efficiency), but this isn't required by
+    the kernel.
+*   `mkfs.erofs` also tries to ensure that the inline data ends in the same
+    disk block as the last byte of the inode metadata (ie: inode header plus
+    xattrs).  This is theoretically not required by the kernel.
+*   A bug present in the kernel before 6.12 meant that this was required for
+    inline symlink targets. This was fixed by `9ed50b8231e3` ("erofs: fix
+    incorrect symlink detection in fast symlink").
+*   In general, when faced with the task of writing out an inode with inline
+    data present, you may need to add padding bytes before the start of the
+    inode in order to ensure that the inline data falls within a single block.
```

Comment:
but the negative impacts of `tail inline` is that
previous in principle you could read `64` (4096/64) inodes at once for each metadata 4k i/o,
now you can only read 1 or 2 inodes if the tail inline is too huge.
It could cause some impacts to `ls` workload (because it just read metadata), so I tend to just inline _small_ data.

Reply:
This is a pretty good point.

Reply:
Keep in mind, though, that for composefs, this is only about directories.  We inline symlinks unconditionally but they're usually very small.  As for regular files, we only inline those when they're less than 32 bytes: otherwise they're handled as redirects to the data layer.

Being about directories it again turns into a question of the cost of a single operation vs the value of reducing the overall amount of IO (because adjacent operations effectively end up prepopulating the cache for future operations when things are more tightly packed).  Imagine the metric is `find /` instead of `ls` for example.

Unlike the question of padding between inodes, though, my gut feeling is that the question could go in the other direction here: wasting more space (in the blocks sections) helps keep the inode table smaller, which helps load it faster and keep it in the cache....

This is again where it would be nice to say something about benchmarks...

Reply:
What's your gut feeling about the appropriate number for inlining directories?  or: at what point do we do another block?

Oh: on that topic there is a slight ambiguity here that I never resolved.  Let's say I'm writing a small directory in 100 bytes and I put it into a single block.  What is the size field?  100 or 4096?  If it was inline it would obviously be 100.  This has a potential impact in knowing where the name of the last entry is terminated (end of file vs scanning for nul).  I'd say 100 but composefs currently does 4096.

Does it matter at all?

Reply:
> Keep in mind, though, that for composefs, this is only about directories. We inline symlinks unconditionally but they're usually very small. As for regular files, we only inline those when they're less than 32 bytes: otherwise they're handled as redirects to the data layer.
> 
> Being about directories it again turns into a question of the cost of a single operation vs the value of reducing the overall amount of IO (because adjacent operations effectively end up prepopulating the cache for future operations when things are more tightly packed). Imagine the metric is `find /` instead of `ls` for example.
> 
> Unlike the question of padding between inodes, though, my gut feeling is that the question could go in the other direction here: wasting more space (in the blocks sections) helps keep the inode table smaller, which helps load it faster and keep it in the cache....
> 
> This is again where it would be nice to say something about benchmarks...

So from the on-disk placement, there is no sliver bullet too. But I guess you could also consider fadvise warm up.

Reply:
> What's your gut feeling about the appropriate number for inlining directories? or: at what point do we do another block?
> 
> Oh: on that topic there is a slight ambiguity here that I never resolved. Let's say I'm writing a small directory in 100 bytes and I put it into a single block. What is the size field? 100 or 4096? If it was inline it would obviously be 100. This has a potential impact in knowing where the name of the last entry is terminated (end of file vs scanning for nul). I'd say 100 but composefs currently does 4096.
> 
> Does it matter at all?

It should be 100, 4096 might work if the trailing filename has `\0` but it's not recommended...